janus_core/helpers/utils.py

```diff
@@ -407,3 +412,50 @@ def track_progress(sequence: Union[Sequence, Iterable], description: str) -> Ite
 
     with progress:
         yield from progress.track(sequence, description=description)
+
+
+def slicelike_to_startstopstep(index: SliceLike) -> StartStopStep:
+    """
+    Standarize `SliceLike`s into tuple of `start`, `stop`, `step`.
+
+    Parameters
+    ----------
+    index : SliceLike
+        `SliceLike` to standardize.
+
+    Returns
+    -------
+    StartStopStep
+        Standardized `SliceLike` as `start`, `stop`, `step` triplet.
+    """
+    if isinstance(index, int):
+        if index == -1:
+            return (index, None, 1)
+        return (index, index + 1, 1)
```

Comment:
Is it obvious why this makes sense?

I'd assumed it would be more like converting `range(index)`, or maybe just that specific value (`start == stop`), but I may be missing something

Reply:
Python stops one early. A slice of `x[1:1]` is a zero-element slice, so to extract a single index we need `x[1:2]`.

For `index == -1`, `x[-1:0]` ie `-1+1`, that is also zero-element.

Reply:
Ah ok so I think I was mixing up a couple of things.

I know Python uses half-open intervals, but for some reason I was thinking part of the point of  our `StartStopStep` was to do more like what we discussed a long time ago for the MD temperature ramp, where users may want to define both end points and we construct the range for them.

I see what you mean though.

I'm not sure the `-1` case works when it's used elsewhere though e.g. trying to use `slicelike_len_for`:

```

In [70]: x = [0,1,2,3,4,5,6,7,8,9,10]
In [71]: sliceable_length = len(x)
In [72]: slc = -1

In [73]: start, stop, step = slicelike_to_startstopstep(slc)
    ...: if stop is None:
    ...:     stop = sliceable_length
    ...: print(len(range(start, stop, step)))
12
```

I may be misusing `slicelike_len_for` though

Reply:
No, that looks to be a bug in `slicelike_len_for` due to the fact it has `-1:sliceable_len` i.e. `-1, 0, ... sliceable_len` == `len+2 == 12`

Reply:
I've updated ``slicelike_len_for``, now ``select_len`` (cf below) and this is also a test case. ``list`` is also handled with it as well as integers